core/deliverservice/bft_client.go

```diff
@@ -0,0 +1,546 @@
+	}
+
+	var numEP int
+	var numRetries int
+	var stopRetries = time.Now().Add(c.reconnectTotalTimeThreshold)
+
+	for !c.shouldStop() {
+		if numEP, err = c.assignReceivers(); err != nil {
+			bftLogger.Debugf("[%s] Exit: error=%s", c.chainID, err)
+			return nil, err
+		}
+
+		c.launchHeaderReceivers()
+
+		response, err = c.receiveBlock()
+		if err == nil {
+			bftLogger.Debugf("[%s] Exit: response=%v", c.chainID, response)
+			return response, nil // the normal return path
+		}
+
+		if stopRetries.Before(time.Now()) {
+			bftLogger.Debugf("[%s] Exit: reconnectTotalTimeThreshold: %s, expired; error: %s",
+				c.chainID, c.reconnectTotalTimeThreshold, errClientReconnectTimeout)
+			return nil, errClientReconnectTimeout
+		}
+
+		c.closeBlockReceiver()
+		numRetries++
+		if numRetries%numEP == 0 { //double the back-off delay on every round of attempts.
+			dur := backOffDuration(2.0, uint(numRetries/numEP), c.minBackoffDelay, c.maxBackoffDelay)
+			bftLogger.Debugf("[%s] Got receive error: %s; going to retry another endpoint in: %s", c.chainID, err, dur)
+			backOffSleep(dur, c.stopChan)
+		} else {
+			bftLogger.Debugf("[%s] Got receive error: %s; going to retry another endpoint now", c.chainID, err)
+		}
+	}
+
+	bftLogger.Debugf("[%s] Exit: %s", c.chainID, errClientClosing.Error())
+	return nil, errClientClosing
+}
+
+func backOffDuration(base float64, exponent uint, minDur, maxDur time.Duration) time.Duration {
+	if base < 1.0 {
+		base = 1.0
+	}
+	if minDur <= 0 {
+		minDur = bftMinBackoffDelay
+	}
+	if maxDur < minDur {
+		maxDur = minDur
+	}
+
+	fDurNano := math.Pow(base, float64(exponent)) * float64(minDur.Nanoseconds())
+	fDurNano = math.Min(fDurNano, float64(maxDur.Nanoseconds()))
+	return time.Duration(fDurNano)
+}
+
+func backOffSleep(backOffDur time.Duration, stopChan <-chan struct{}) {
+	select {
+	case <-time.After(backOffDur):
+	case <-stopChan:
+	}
+}
+
+// Check block reception progress relative to header reception progress.
+// If the orderer associated with the block receiver is suspected of censorship, replace it with another orderer.
+func (c *bftDeliveryClient) monitor() {
+	bftLogger.Debugf("[%s] Entry", c.chainID)
+
+	ticker := time.NewTicker(bftBlockCensorshipTimeout / 100)
+	for !c.shouldStop() {
+		if suspicion := c.detectBlockCensorship(); suspicion {
+			c.closeBlockReceiver()
+			c.lastBlockTime = time.Now()
+		}
+
+		select {
+		case <-ticker.C:
+		case <-c.stopChan:
+		}
+	}
+	ticker.Stop()
+
+	bftLogger.Debugf("[%s] Exit", c.chainID)
+}
+
+func (c *bftDeliveryClient) detectBlockCensorship() bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.blockReceiver == nil {
+		return false
+	}
+
+	now := time.Now()
+	suspicionThreshold := c.lastBlockTime.Add(c.blockCensorshipTimeout)
+	if now.Before(suspicionThreshold) {
+		return false
+	}
+
+	var numAhead int
+	for ep, hRcv := range c.headerReceivers {
+		blockNum, _, err := hRcv.LastBlockNum()
+		if err != nil {
```

Comment:
we should log the error

Reply:
It is logged as Warning in the called function, LastBlockNum()